### PR TITLE
New tag map class for both unknown and extension values.

### PR DESCRIPTION
**Pull request by swankjesse**
This isn't wired in yet, but I wanted to get this reviewed
independently. The implementation is interesting!

### CODE REVIEW

File: wire-runtime/src/main/java/com/squareup/wire/NewTagMap.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ * <p>Tags are stored in the order that they were added. Values that are semantically equal may be
+ * unequal according to this class because this implementation doesn't implement a strong order-
+ * independent equals. This is unlikely to be a problem in practice because encoders typically use a
+ * predictable encoding order.
+ */
+final class NewTagMap {
+  private int size = 0;
+
+  /** Tags. Defined only for [0..size). */
+  private int[] tags = new int[8];
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/NewTagMap.java]: Assuming instances will be lazily initialized

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/NewTagMap.java]: For now.

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/NewTagMap.java]: (it's mutable, so anything else is a hassle!)

**Comment by JakeWharton** [comment]
Cool tests. LGTM!

**Pull request merged**